x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/bulk/TransportBulkShardOperationsAction.java

```diff
@@ -151,20 +181,38 @@ protected BulkShardOperationsResponse newResponseInstance() {
     /**
      * Custom write result to include global checkpoint after ops have been replicated.
      */
-    static class CcrWritePrimaryResult extends WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> {
+    static final class CcrWritePrimaryResult extends WritePrimaryResult<BulkShardOperationsRequest, BulkShardOperationsResponse> {
+        final long waitingForGlobalCheckpoint;
 
-        CcrWritePrimaryResult(BulkShardOperationsRequest request, Translog.Location location, IndexShard primary, Logger logger) {
+        CcrWritePrimaryResult(BulkShardOperationsRequest request, Translog.Location location, IndexShard primary,
+                              long waitingForGlobalCheckpoint, Logger logger) {
             super(request, new BulkShardOperationsResponse(), location, null, primary, logger);
+            this.waitingForGlobalCheckpoint = waitingForGlobalCheckpoint;
         }
 
         @Override
         public synchronized void respond(ActionListener<BulkShardOperationsResponse> listener) {
-            final BulkShardOperationsResponse response = finalResponseIfSuccessful;
-            final SeqNoStats seqNoStats = primary.seqNoStats();
-            // return a fresh global checkpoint after the operations have been replicated for the shard follow task
-            response.setGlobalCheckpoint(seqNoStats.getGlobalCheckpoint());
-            response.setMaxSeqNo(seqNoStats.getMaxSeqNo());
-            listener.onResponse(response);
+            final Runnable fillResponse = () -> {
+                final BulkShardOperationsResponse response = finalResponseIfSuccessful;
+                final SeqNoStats seqNoStats = primary.seqNoStats();
+                // return a fresh global checkpoint after the operations have been replicated for the shard follow task
+                response.setGlobalCheckpoint(seqNoStats.getGlobalCheckpoint());
+                response.setMaxSeqNo(seqNoStats.getMaxSeqNo());
+            };
+            if (waitingForGlobalCheckpoint != SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                primary.addGlobalCheckpointListener(waitingForGlobalCheckpoint, (gcp, e) -> {
+                    if (e != null) {
+                        listener.onFailure(e);
+                    } else {
+                        assert waitingForGlobalCheckpoint <= gcp : waitingForGlobalCheckpoint + " > " + gcp;
+                        fillResponse.run();
```

Comment:
fillResponse can throw an already closed exception. We should make sure we deal with exceptions here correctly

Reply:
Maybe warp the listener using ActionListener#wrap which does the write things and will simplify the code here too.